File: scala/acronym/src/main/scala/Acronym.scala
object Acronym {

  def abbreviate(phrase: String): String = {
    def acronymChar(x: Char, y: Char): Option[Char] =
      if (!x.isLetter && y.isLetter) Some(y.toUpper)
      else if (!x.isUpper && y.isUpper) Some(y)
      else None

    val charPairs = ' ' +: phrase zip phrase

    charPairs flatMap (acronymChar _).tupled.andThen(_.toList) mkString
  }
}

File: haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char (toUpper, isUpper, isSpace, isLetter)
import Data.Maybe (mapMaybe)

abbreviate :: String -> String
-- abbreviate = mapMaybe acronymChar . (zip =<< (' ' :)) -- by pointfree.io
abbreviate xs = mapMaybe acronymChar charPairs
  where
    charPairs = zip (' ':xs) xs
    acronymChar (x, y)
      | (not . isLetter) x && isLetter y = Just $ toUpper y
      | (not . isUpper) x  && isUpper y  = Just y
      | otherwise = Nothing
